client/src/worker_registry/mod.rs

```diff
@@ -0,0 +1,288 @@
+//! This module enables the tracking of workers that are associated with a client instance.
+//! This is needed to implement Eager Workflow Start, a latency optimization in which the client,
+//!  after reserving a slot, directly forwards a WFT to a local worker.
+
+use rand::{seq::SliceRandom, thread_rng};
+use std::{collections::HashMap, fmt::Debug};
+use temporal_sdk_core_protos::temporal::api::workflowservice::v1::PollWorkflowTaskQueueResponse;
+
+/// This trait wraps a worker, which provides WFT processing slots.
+#[cfg_attr(test, mockall::automock)]
+pub trait SlotProvider: Send + Sync + Debug {
+    /// A unique identifier for the worker.
+    fn uuid(&self) -> String;
+    /// The namespace for the WFTs that it can process.
+    fn namespace(&self) -> String;
+    /// The task queue this provider listens to.
+    fn task_queue(&self) -> String;
+    /// Try to reserve a slot with matching namespace and task queue.
+    fn try_reserve_wft_slot(&self) -> Option<Box<dyn Slot>>;
+}
+
+/// This trait represents a slot reserved for processing a WFT by a worker.
+#[cfg_attr(test, mockall::automock)]
+pub trait Slot: Send + Sync {
+    /// Consumes this slot by dispatching a WFT to its worker. This can only be called once.
+    fn schedule_wft(&mut self, task: PollWorkflowTaskQueueResponse) -> Result<(), anyhow::Error>;
```

Comment:
The hint is already in the docstring - this should take `self` rather than `&mut self`, since it "consumes" the slot.

Reply:
Yes, as I mentioned above, I tried to make it `self` but is a `Dyn Slot`, and Rust doesn't let me move it since it does not know the size. Is there a way to do this with Dyn?

Reply:
Consuming a Box<Slot> did the trick...